File: scala/parsers/src/main/scala/jp/co/teruuu/parser/common/StopWithEscape.scala
package jp.co.teruuu.parser.common

class StopWithEscape(stop: String, escapes: Map[String, String]) extends Parser[String] {
  assert(stop.length > 0)

  override def parse(input: String): ParseResult[String] = {
    var next = input
    var continueFlg = true
    var ret = ""
    while(continueFlg) {
      escapes.foreach{case (key, value) => {
        if(next.startsWith(key)) {
          ret += value
          next = next.substring(key.length)
        }
      }}
      if(next.length > 0 && !next.startsWith(stop)) {
        ret += next(0)
        next = next.substring(1)
      } else {
        continueFlg = false
      }
    }
    ParseOk(ret, next)
  }
}

File: typescript/parser/src/parser/common/stop-with-escape.ts
import { Parser } from './parser';
import { ParseResult, ParseSuccess, ParseFailer } from './parser-result';

export class StopWithEscape implements Parser<string> {
	stop: string;
	escapes: Map<string, string>;
	constructor(stop: string, escapes: Map<string, string>) {
		this.stop = stop;
		this.escapes = escapes;
	}

	parse(input: string): ParseResult<String> {
		let next = input
		let result = ""
		let continueFlg = true
		while(continueFlg) {
			this.escapes.forEach((value,key) => {
				if(next.startsWith(key)) {
					result += value
					next = next.substring(key.length)
				}
			})
			if(next.length > 0 && !next.startsWith(this.stop)) {
				result += next.substring(0, 1)
				next = next.substring(1)
			} else {
				continueFlg = false
			}
		}
		return new ParseSuccess(result, next);
	}
}
